### DISCUSSION TITLE
Adding non-casadi code to function solved in parallel

### DISCUSSION THREAD

**Discussion by akb427** [casadi/casadi#3880] [Q&A] [answered]
I am trying to solve a decentralized optimization problem. I create 4 optifunctions, which each contain a call to a custom numerical-only code in call (created following [these](https://github.com/casadi/casadi/wiki/FAQ%3A-how-to-add-non-CasADi-code-into-a-CasADi-optimization%3F) instructions). Then I try to solve these functions in parallel using the parfor loop in matlab. However, I get the following error message. Is there anyway to work around this? The code works when not solved in parallel.
 
` Warning: Serializing of CasADi Function failed:Error using casadi.Function.deserialize (line 5828)
.../casadi/core/function_internal.cpp:4037: Assertion "it!=FunctionInternal::deserialize_map.end()" failed:
FunctionInternal::deserialize: not found 'CallbackInternal'
Error in casadi.Function.loadobj (line 5834)
obj = casadi.Function.deserialize(s.serialization);
Error in parallel.internal.pool.optionallyDeserialize (line 7)
out = distcompdeserialize(data)
In casadi.Function.loadobj (line 5839)
In parallel.internal.pool.optionallyDeserialize (line 7)
Warning: While loading an object of class 'casadi.Function':
Output argument "obj" (and possibly others) not assigned a value in the execution with "casadi.Function.loadobj" function.
In parallel.internal.pool.optionallyDeserialize (line 7)`

**Comment by jgillis** [comment, marked as answer]
3.6.7 contains an experimental alternative that would work in this situation: matlab_ipc. Unfortunately haven't made a public example yet.

**Reply by akb427** [reply]
My custom code is designed to map an values of a casadi variable to a single limit if over a threshhold value. Essentially, I am trying to create the conditional constraint that 0<x<10^3 or x==10^5, as I am trying to create a model of a valve, where 10^5 is equivalent to infinity or closed. Is there a better way to do this that does not involve a numeric code call?

**Reply by jgillis** [reply]
Can you write your logic here? Probably implementable in casadi expressions directly

**Reply by akb427** [reply]
Yes it is fairly simple; I was also considering the big-M approach, but as the problem is already very nonlinear and numerically difficult, I thought that might further escalate the issues.

```
% In the main code: 
zeta_u = opti_flow.variable(num.user, num.seg);
opti_flow.subject_to(0<zeta_u(:)<2*10^3)
zeta_u_lim = mySaturation(zeta_u);

% mySaturation (the custom function)
classdef MySaturation < casadi.Callback
    methods
        function self = MySaturation()
            self@casadi.Callback();
            construct(self, 'Saturation', struct('enable_fd',true));
        end

        function [results] = eval(self, arg)
          arg = full(arg{1});
          res = arg;
          res(arg>10^3)=5*10^5;
          results = {res};
        end
    end
end

**Reply by jgillis** [reply]
res = if_else(arg>10^3,5*10^5,arg)
